sway-core/src/semantic_analysis/ast_node/declaration/struct.rs

```diff
@@ -57,8 +60,43 @@ impl ty::TyStructDecl {
 
             // type check the fields
             let mut new_fields = vec![];
+            let mut encountered_non_indexed_field = false;
             for field in fields.into_iter() {
-                new_fields.push(ty::TyStructField::type_check(handler, ctx.by_ref(), field)?);
+                let ty_field = ty::TyStructField::type_check(handler, ctx.by_ref(), field)?;
+                if ty_field.attributes.indexed().is_some() && attributes.event().is_none() {
+                    return Err(
+                        handler.emit_err(CompileError::IndexedFieldInNonEventStruct {
+                            field_name: ty_field.name.into(),
+                            struct_name: name.into(),
+                        }),
+                    );
+                }
+
+                let abi_size_hint = ctx
+                    .engines()
+                    .te()
+                    .get(ty_field.type_argument.type_id)
+                    .abi_encode_size_hint(ctx.engines());
+                if ty_field.attributes.indexed().is_some() {
```

Comment:
nit: move inside the block to avoid unneccesary compuation since it's not used outside?

```suggestion
                if ty_field.attributes.indexed().is_some() {
                    let abi_size_hint = ctx
                        .engines()
                        .te()
                        .get(ty_field.type_argument.type_id)
                        .abi_encode_size_hint(ctx.engines());
```